Computed and propagated the value of OldEndpoints field when merging …

akstron opened the pull request:
…information from remote clusters.

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

- Updated `mergeServiceUpdateLocked` and `mergeExternalServiceDeleteLocked` to calculate and propagate oldEndpoints.

Fixes: #26176 

```release-note
Computed and propagated the value of OldEndpoints field when merging remote cluster information.
```
Signed-off-by: Alok Kumar Singh <alokaks601@gmail.com>

giorio94 commented on pkg/k8s/service_cache.go:
```diff
@@ -550,10 +550,13 @@ func (s *ServiceCache) mergeServiceUpdateLocked(service *serviceStore.ClusterSer
 		s.externalEndpoints[id] = externalEndpoints
 	}
 
+	var oldEPs *Endpoints
```
I'd suggest to use `s.correlateEndpoints(id)` to compute the old endpoints. This ensures that we actually propagate all old endpoints, and not only the ones associated with the given remote cluster, so that the difference between new and old endpoints corresponds to the ones that have been removed. The same commend applies to the other occurrence.

akstron replied:
Got it. Shouldn't it be done in case of local cluster as well? As I can see this is not done in case of local cluster update and delete of earlier pr.

giorio94 replied:
Yes, I would uniform that as well. It would also be cool extending the existing tests to assert the correctness of `OldEndpoints`.

giorio94 reviewed (changes requested):
Thanks @akstron for your contribution!

In addition to the inline comment, could you also please extend the commit message with the rationale behind this change?

akstron committed:
Computing and propagating OldEndpoints

OldEndpoints field was added to fix bugs in toService policies. The field was earlier not computed when
merging information from remote clusters.

This commit aims to calculate and propagate the OldEndpoints field value when handling remote cluster.
It is mainly performed for consistency.

Signed-off-by: Alok Kumar Singh <alokaks601@gmail.com>
